R/shuffle_card.R

```diff
@@ -0,0 +1,301 @@
+#' Shuffle `cards`
+#'
+#' @description `r lifecycle::badge('experimental')`\cr
+#'
+#' This function ingests an ARD object and shuffles the information to prepare
+#' for analysis. Helpful for streamlining across multiple ARDs.
+#'
+#' @param x an ARD data frame of class 'card'
+#' @param by Grouping variable(s) used in calculations. Defaults to `NULL`. If
+#'   available (i.e. if `x` comes from a stacking function), `attributes(x)$by`
+#'   will be used instead of `by`.
+#' @param trim logical representing whether or not to trim away `fmt_fun`,
+#'   `error`, and `warning` columns
+#' @param fill_overall scalar to fill missing grouping or variable levels. If a
+#'   character is passed, then it is processed with [glue::glue()] where the
+#'   colname element is available to inject into the string,
+#'   e.g. `Overall {colname}` may resolve to `"Overall AGE"` for an `AGE`
+#'   column. Default is `Overall {colname}`. If `NA` then no fill will occur.
+#' @param fill_hierarchical_overall scalar to fill variable levels for overall
+#'   hierarchical calculations. If a character is passed, then it is processed
+#'   with [glue::glue()] where the colname element is available to inject into
+#'   the string, e.g. `Any {colname}` may resolve to `"Any AESOC"` for an
+#'   `AESOC` column. Default is `Any {colname}`. If `NA` then no fill will
+#'   occur.
+#'
+#' @return a tibble
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' cards::bind_ard(
+#'   cards::ard_categorical(cards::ADSL, by = "ARM", variables = "AGEGR1"),
+#'   cards::ard_categorical(cards::ADSL, variables = "ARM")
+#' ) |>
+#'   shuffle_card()
+shuffle_card <- function(x,
+                         by = NULL,
+                         trim = TRUE,
+                         fill_overall = "Overall {colname}",
+                         fill_hierarchical_overall = "Any {colname}") {
+
+  if (!inherits(x, "card")) {
+    cli::cli_abort(
+      "{.arg x} argument must be class {.cls card}, not {.obj_type_friendly {x}}",
+    env = rlang::caller_env())
+  }
+  if (!inherits(trim, "logical")) {
+    cli::cli_abort(
+      "{.arg trim} argument must be class {.cls logical}}, not {.obj_type_friendly {trim}}",
+      env = rlang::caller_env())
+  }
+
+  ard_args <- attributes(x)$args
+  by <- .process_by(x, by)
+
+  # make sure columns are in order & add index for retaining order
+  dat_cards <- x |>
+    cards::tidy_ard_column_order() |>
+    cards::tidy_ard_row_order() |>
```

Comment:
i think we should make sorting optional, since cards now has hierarchical sorting utilities.